File: frontend/src/utils/response.ts
// 对应 response.py 的结构

export type SystemMessageType = 'info' | 'warning' | 'success' | 'error';

import { AgentType } from './enum';


export interface BaseMessage {
  id: string;
  msg_type: 'system' | 'agent' | 'user' | 'tool';
  content?: string | null;
}

export interface ToolMessage extends BaseMessage {
  msg_type: 'tool';
  tool_name: 'execute_code' | 'search_scholar';
  input: any;
  output: string[] | OutputItem[] | null;
}

export interface SystemMessage extends BaseMessage {
  msg_type: 'system';
  type: SystemMessageType;
}

export interface UserMessage extends BaseMessage {
  msg_type: 'user';
}

export interface AgentMessage extends BaseMessage {
  msg_type: 'agent';
  agent_type: AgentType;
}

export interface ModelerMessage extends AgentMessage {
  agent_type: AgentType.MODELER;
}

export interface CoordinatorMessage extends AgentMessage {
  agent_type: AgentType.COORDINATOR;
}


// 代码执行结果类型
export type ExecutionFormat = 
  | 'text'
  | 'html'
  | 'markdown'
  | 'png'
  | 'jpeg'
  | 'svg'
  | 'pdf'
  | 'latex'
  | 'json'
  | 'javascript';

export interface BaseCodeExecution {
  res_type: 'stdout' | 'stderr' | 'result' | 'error';
  msg?: string;
}

export interface StdOutExecution extends BaseCodeExecution {
  res_type: 'stdout';
}

export interface StdErrExecution extends BaseCodeExecution {
  res_type: 'stderr';
}

export interface ResultExecution extends BaseCodeExecution {
  res_type: 'result';
  format: ExecutionFormat;
}

export interface ErrorExecution extends BaseCodeExecution {
  res_type: 'error';
  name: string;
  value: string;
  traceback: string;
}

export type OutputItem = StdOutExecution | StdErrExecution | ResultExecution | ErrorExecution;

export interface ScholarMessage extends ToolMessage {
  tool_name: 'search_scholar';
  input: {};
  output: string[];
}

export interface InterpreterMessage extends ToolMessage {
  tool_name: 'execute_code';
  input: {
    code: string;
  } | null;
  output: OutputItem[] | null;
}


export interface CoderMessage extends AgentMessage {
  agent_type: AgentType.CODER;
}

export interface WriterMessage extends AgentMessage {
  agent_type: AgentType.WRITER;
  sub_title?: string;
}

export type Message = SystemMessage | UserMessage | CoderMessage | WriterMessage | ModelerMessage | CoordinatorMessage | ToolMessage;

File: backend/app/schemas/response.py
from typing import Literal, Union
from app.schemas.enums import AgentType
from pydantic import BaseModel, Field
from uuid import uuid4


class Message(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    msg_type: Literal[
        "system", "agent", "user", "tool"
    ]  # system msg | agent message | user message | tool message
    content: str | None = None


class ToolMessage(Message):
    msg_type: str = "tool"
    tool_name: Literal["execute_code", "search_scholar"]
    input: dict
    output: list


class SystemMessage(Message):
    msg_type: str = "system"
    type: Literal["info", "warning", "success", "error"] = "info"


class UserMessage(Message):
    msg_type: str = "user"


class AgentMessage(Message):
    msg_type: str = "agent"
    agent_type: AgentType  # CoordinatorAgent | ModelerAgent | CoderAgent | WriterAgent


class ModelerMessage(AgentMessage):
    agent_type: AgentType = AgentType.MODELER


class CoordinatorMessage(AgentMessage):
    agent_type: AgentType = AgentType.COORDINATOR


class CodeExecution(BaseModel):
    res_type: Literal["stdout", "stderr", "result", "error"]
    msg: str | None = None


class StdOutModel(CodeExecution):
    res_type: str = "stdout"


class StdErrModel(CodeExecution):
    res_type: str = "stderr"


class ResultModel(CodeExecution):
    res_type: str = "result"
    format: Literal[
        "text",
        "html",
        "markdown",
        "png",
        "jpeg",
        "svg",
        "pdf",
        "latex",
        "json",
        "javascript",
    ]


class ErrorModel(CodeExecution):
    res_type: str = "error"
    name: str
    value: str
    traceback: str


# 代码执行结果类型
OutputItem = Union[StdOutModel, StdErrModel, ResultModel, ErrorModel]


class ScholarMessage(ToolMessage):
    tool_name: str = "search_scholar"
    input: dict | None = None  # query
    output: list[str] | None = None  # cites


class InterpreterMessage(ToolMessage):
    tool_name: str = "execute_code"
    input: dict | None = None  # code
    output: list[OutputItem] | None = None  # code_results


# 1. 只带 code
# 2. 只带 code result
class CoderMessage(AgentMessage):
    agent_type: AgentType = AgentType.CODER


class WriterMessage(AgentMessage):
    agent_type: AgentType = AgentType.WRITER
    sub_title: str | None = None


# 所有可能的消息类型
MessageType = Union[
    SystemMessage,
    UserMessage,
    ModelerMessage,
    CoderMessage,
    WriterMessage,
    CoordinatorMessage,
]
